src/adapter/src/coord/sequencer/cluster.rs

```diff
@@ -671,30 +676,37 @@ impl Coordinator {
                     name,
                     &compute,
                     new_size,
-                    &mut ops,
+                    &mut create_ops,
                     // AVAILABILITY ZONES hasn't changed, so existing replicas don't need to be
                     // rescheduled.
                     Some(new_availability_zones.as_ref()),
                     *new_disk,
+                    new_replica_start_pending,
                     owner_id,
                     reason.clone(),
                 )?;
                 create_cluster_replicas.push((cluster_id, id))
             }
         }
 
-        let variant = ClusterVariant::Managed(new_config);
-        ops.push(catalog::Op::UpdateClusterConfig {
-            id: cluster_id,
-            name,
-            config: ClusterConfig { variant },
-        });
-
-        self.catalog_transact(session, ops).await?;
+        // we only finalize if we created something and the thing we created
+        // started as pending
+        let mut needs_finalization = true;
+        if create_cluster_replicas.is_empty() || !new_replica_start_pending {
+            let variant = ClusterVariant::Managed(new_config);
+            drop_ops.push(catalog::Op::UpdateClusterConfig {
+                id: cluster_id,
+                name,
+                config: ClusterConfig { variant },
+            });
+            self.catalog_transact(session, drop_ops).await?;
+            needs_finalization = false;
+        }
+        self.catalog_transact(session, create_ops.clone()).await?;
         for (cluster_id, replica_id) in create_cluster_replicas {
             self.create_cluster_replica(cluster_id, replica_id).await;
         }
-        Ok(())
+        Ok(needs_finalization)
```

Comment:
Returning a requirement to a caller is somewhat error prone as a boolean. Is there a better type we can return instead? Or at least name and document this as something that happened instead of something the caller must do, for example it could instead be named `pending_replicas_created` then it's up to the caller to decide what to do with that information.

Another possibility is to return an `Option<Duration>` which is extracted from the strategy. That sounds enormously better to me because: it is much more clear what the caller needs to do with that return value (bools are bad at function boundaries), and will also start to cause compile errors when the other strategy is implemented (wait for thing to become ready), which is good! That may lead us to create some new enum that describes the wait action that should occur, which the caller must then also handle in order to compile. If this were a boolean and the other wait conditions were added, you'd have to figure out where they needed to go instead of the compiler telling you exactly where they were needed.

Reply:
Yeah thanks for the suggestions! I didn't feel great about this and had intended to revisit! I had considered a few things, and nearly landed an a `NeedsFinalizaion` or `RequiredAction` enum, I might go back to that.. I think I might prefer an enum here so it's less confusing to think about until caught up.